lhotse/features/base.py

```diff
@@ -0,0 +1,411 @@
+from abc import ABCMeta, abstractmethod
+from concurrent.futures.process import ProcessPoolExecutor
+from dataclasses import is_dataclass, asdict, dataclass, field
+from functools import partial
+from itertools import chain
+from math import isclose
+from pathlib import Path
+from typing import Optional, Any, List, Iterable, Type
+
+import lilcom
+import numpy as np
+import torch
+
+from lhotse.audio import Recording
+from lhotse.supervision import SupervisionSegment
+from lhotse.utils import Seconds, Pathlike, load_yaml, save_to_yaml, uuid4
+
+
+class FeatureExtractor(metaclass=ABCMeta):
+    """
+    The base class for all feature extractors in Lhotse.
+    It is initialized with a config object, specific to a particular feature extraction method.
+    The config is expected to be a dataclass so that it can be easily serialized.
+
+    All derived feature extractors must implement at least the following:
+    - a ``name`` class attribute (how are these features called, e.g. 'mfcc')
+    - a ``config_type`` class attribute that points to the configuration dataclass type
+    - the ``extract`` method,
+    - the ``frame_shift`` property.
+
+    Feature extractors that support feature-domain mixing should additionally specify two static methods:
+    - ``compute_energy``, and
+    - ``mix``.
+
+    It is expected that when implemented, ``compute_energy`` will never return zero.
+    """
+    name = None
+    config_type = None
+
+    def __init__(self, config: Optional[Any] = None):
+        if config is None:
+            config = self.config_type()
+        assert is_dataclass(config), "The feature configuration object must be a dataclass."
+        self.config = config
+
+    @abstractmethod
+    def extract(self, samples: np.ndarray, sampling_rate: int) -> np.ndarray: ...
+
+    @property
+    @abstractmethod
+    def frame_shift(self) -> Seconds: ...
+
+    @staticmethod
+    def mix(features_a: np.ndarray, features_b: np.ndarray, gain_b: float) -> np.ndarray:
```

Comment:
would be nice to have a doc-string saying e.g. what gain_b means.  E.g. is it in db?  (Clearer variable name, if so)...

Reply:
While writing this doc-string, I noticed that when mixing audio, we were using a gain value estimated for energy/power mixing, and not amplitude mixing. I added an `sqrt()` call and a comment in that place.